File: eval/compute_significance_stats.py
# -*- coding: utf-8 -*-
"""Compute paired significance statistics for repeated-run evaluations.

Inputs are the 10 evaluation folders under:
  repaet run eval/best_baseseed{seed}
  repaet run eval/best_fullseed{seed}

This script is used for the revised manuscript's five-seed statistical
validation. It pools paired seed-image observations after checking that the
baseline and full-model prediction files have the same image paths and labels.

Outputs:
  repaet run eval/significance_stats/significance_summary.csv
  repaet run eval/significance_stats/significance_summary.md
  repaet run eval/significance_stats/significance_per_seed.csv
  repaet run eval/significance_stats/mean_class_delta_f1.csv
"""

from __future__ import annotations

import argparse
import json
from pathlib import Path

import numpy as np
import pandas as pd
from scipy.stats import binomtest, wilcoxon


NUM_CLASSES = 241


def read_predictions(eval_root: Path, method: str, seed: int) -> pd.DataFrame:
    prefix = "best_baseseed" if method == "baseline" else "best_fullseed"
    path = eval_root / f"{prefix}{seed}" / "predictions.csv"
    df = pd.read_csv(path)
    df["top5_hit"] = [
        int(int(y) in {int(x) for x in str(ids).split(";")})
        for y, ids in zip(df["true_id"], df["top5_ids"])
    ]
    return df


def read_class_f1(eval_root: Path, method: str, seed: int) -> np.ndarray:
    prefix = "best_baseseed" if method == "baseline" else "best_fullseed"
    path = eval_root / f"{prefix}{seed}" / "classification_report.csv"
    df = pd.read_csv(path, index_col=0)
    values = []
    for i in range(NUM_CLASSES):
        key = f"{i:03d}"
        values.append(float(df.loc[key, "f1-score"]))
    return np.asarray(values, dtype=float)


def f1_scores_fast(y_true: np.ndarray, pred: np.ndarray) -> tuple[float, float]:
    support = np.bincount(y_true, minlength=NUM_CLASSES).astype(float)
    pred_count = np.bincount(pred, minlength=NUM_CLASSES).astype(float)
    correct = pred == y_true
    tp = np.bincount(y_true[correct], minlength=NUM_CLASSES).astype(float)
    fp = pred_count - tp
    fn = support - tp
    denom = 2 * tp + fp + fn
    f1 = np.divide(2 * tp, denom, out=np.zeros_like(tp, dtype=float), where=denom > 0)
    macro_f1 = float(np.mean(f1))
    weighted_f1 = float(np.sum(f1 * support) / np.sum(support))
    return macro_f1, weighted_f1


def paired_delta(y_true, base_pred, full_pred, base_top5, full_top5) -> dict:
    base_macro, base_weighted = f1_scores_fast(y_true, base_pred)
    full_macro, full_weighted = f1_scores_fast(y_true, full_pred)
    return {
        "top1": float(np.mean(full_pred == y_true) - np.mean(base_pred == y_true)),
        "top5": float(np.mean(full_top5) - np.mean(base_top5)),
        "macro_f1": float(full_macro - base_macro),
        "weighted_f1": float(full_weighted - base_weighted),
    }


def paired_bootstrap_ci(y_true, base_pred, full_pred, base_top5, full_top5, n_boot: int, seed: int) -> dict:
    rng = np.random.default_rng(seed)
    n = len(y_true)
    samples = {k: [] for k in ["top1", "top5", "macro_f1", "weighted_f1"]}
    for _ in range(n_boot):
        idx = rng.integers(0, n, size=n)
        d = paired_delta(
            y_true[idx],
            base_pred[idx],
            full_pred[idx],
            base_top5[idx],
            full_top5[idx],
        )
        for k, v in d.items():
            samples[k].append(v)
    return {k: tuple(np.percentile(v, [2.5, 97.5])) for k, v in samples.items()}


def mcnemar_exact(y_true, base_pred, full_pred) -> dict:
    base_correct = base_pred == y_true
    full_correct = full_pred == y_true
    b01 = int(np.sum((~base_correct) & full_correct))
    b10 = int(np.sum(base_correct & (~full_correct)))
    n = b01 + b10
    p = float(binomtest(min(b01, b10), n=n, p=0.5, alternative="two-sided").pvalue) if n else 1.0
    return {
        "baseline_wrong_full_correct": b01,
        "baseline_correct_full_wrong": b10,
        "discordant": n,
        "p_value": p,
    }


def fmt_p(p: float) -> str:
    return f"{p:.3e}" if p < 1e-4 else f"{p:.6f}"


def fmt_pp(x: float) -> str:
    return f"{x * 100:.3f}"


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--eval-root", default="repaet run eval")
    parser.add_argument("--seeds", default="1,25,42,50,100")
    parser.add_argument("--n-boot", type=int, default=2000)
    parser.add_argument("--out-dir", default=None)
    args = parser.parse_args()

    root = Path(__file__).resolve().parents[1]
    eval_root = (root / args.eval_root).resolve()
    out_dir = Path(args.out_dir) if args.out_dir else eval_root / "significance_stats"
    out_dir.mkdir(parents=True, exist_ok=True)
    seeds = [int(x) for x in args.seeds.split(",") if x.strip()]

    pooled_true = []
    pooled_base_pred = []
    pooled_full_pred = []
    pooled_base_top5 = []
    pooled_full_top5 = []
    per_seed_rows = []
    class_delta_f1 = []

    for seed in seeds:
        base = read_predictions(eval_root, "baseline", seed)
        full = read_predictions(eval_root, "full", seed)
        if len(base) != len(full):
            raise ValueError(f"Prediction length mismatch for seed {seed}: {len(base)} vs {len(full)}")
        if not (base["path"].to_numpy() == full["path"].to_numpy()).all():
            raise ValueError(f"Prediction path mismatch for seed {seed}")
        if not (base["true_id"].to_numpy() == full["true_id"].to_numpy()).all():
            raise ValueError(f"Prediction label mismatch for seed {seed}")

        y = base["true_id"].to_numpy(dtype=int)
        bp = base["pred_id"].to_numpy(dtype=int)
        fp = full["pred_id"].to_numpy(dtype=int)
        b5 = base["top5_hit"].to_numpy(dtype=int)
        f5 = full["top5_hit"].to_numpy(dtype=int)

        delta = paired_delta(y, bp, fp, b5, f5)
        mc = mcnemar_exact(y, bp, fp)
        base_f1 = read_class_f1(eval_root, "baseline", seed)
        full_f1 = read_class_f1(eval_root, "full", seed)
        diff_f1 = full_f1 - base_f1
        wx = wilcoxon(diff_f1, alternative="greater", zero_method="wilcox", correction=False, method="auto")

        per_seed_rows.append({
            "seed": seed,
            "n_images": len(y),
            **{f"delta_{k}": v for k, v in delta.items()},
            "mcnemar_b01": mc["baseline_wrong_full_correct"],
            "mcnemar_b10": mc["baseline_correct_full_wrong"],
            "mcnemar_p": mc["p_value"],
            "wilcoxon_W": float(wx.statistic),
            "wilcoxon_p_greater": float(wx.pvalue),
            "mean_delta_f1": float(np.mean(diff_f1)),
            "positive_classes": int(np.sum(diff_f1 > 0)),
            "negative_classes": int(np.sum(diff_f1 < 0)),
            "zero_classes": int(np.sum(diff_f1 == 0)),
        })

        pooled_true.append(y)
        pooled_base_pred.append(bp)
        pooled_full_pred.append(fp)
        pooled_base_top5.append(b5)
        pooled_full_top5.append(f5)
        class_delta_f1.append(diff_f1)

    y = np.concatenate(pooled_true)
    bp = np.concatenate(pooled_base_pred)
    fp = np.concatenate(pooled_full_pred)
    b5 = np.concatenate(pooled_base_top5)
    f5 = np.concatenate(pooled_full_top5)

    delta = paired_delta(y, bp, fp, b5, f5)
    ci = paired_bootstrap_ci(y, bp, fp, b5, f5, n_boot=args.n_boot, seed=20260617)
    mc = mcnemar_exact(y, bp, fp)
    mean_class_delta = np.mean(np.vstack(class_delta_f1), axis=0)
    wx = wilcoxon(mean_class_delta, alternative="greater", zero_method="wilcox", correction=False, method="auto")

    summary = {
        "comparison": "Full model vs Baseline",
        "seeds": ",".join(map(str, seeds)),
        "n_seed_image_pairs": int(len(y)),
        "mcnemar_b01_baseline_wrong_full_correct": mc["baseline_wrong_full_correct"],
        "mcnemar_b10_baseline_correct_full_wrong": mc["baseline_correct_full_wrong"],
        "mcnemar_discordant": mc["discordant"],
        "mcnemar_p_two_sided_exact": mc["p_value"],
        "delta_top1": delta["top1"],
        "delta_top1_ci_low": ci["top1"][0],
        "delta_top1_ci_high": ci["top1"][1],
        "delta_top5": delta["top5"],
        "delta_top5_ci_low": ci["top5"][0],
        "delta_top5_ci_high": ci["top5"][1],
        "delta_macro_f1": delta["macro_f1"],
        "delta_macro_f1_ci_low": ci["macro_f1"][0],
        "delta_macro_f1_ci_high": ci["macro_f1"][1],
        "delta_weighted_f1": delta["weighted_f1"],
        "delta_weighted_f1_ci_low": ci["weighted_f1"][0],
        "delta_weighted_f1_ci_high": ci["weighted_f1"][1],
        "wilcoxon_W": float(wx.statistic),
        "wilcoxon_p_greater": float(wx.pvalue),
        "wilcoxon_mean_delta_f1": float(np.mean(mean_class_delta)),
        "wilcoxon_positive_classes": int(np.sum(mean_class_delta > 0)),
        "wilcoxon_negative_classes": int(np.sum(mean_class_delta < 0)),
        "wilcoxon_zero_classes": int(np.sum(mean_class_delta == 0)),
    }

    pd.DataFrame([summary]).to_csv(out_dir / "significance_summary.csv", index=False, encoding="utf-8-sig")
    pd.DataFrame(per_seed_rows).to_csv(out_dir / "significance_per_seed.csv", index=False, encoding="utf-8-sig")
    pd.DataFrame({
        "class_id": np.arange(NUM_CLASSES),
        "mean_delta_f1": mean_class_delta,
    }).to_csv(out_dir / "mean_class_delta_f1.csv", index=False, encoding="utf-8-sig")
    (out_dir / "significance_summary.json").write_text(
        json.dumps(summary, ensure_ascii=False, indent=2), encoding="utf-8"
    )

    md = [
        "| Analysis | Statistic | Result |",
        "|---|---:|---:|",
        f"| McNemar's test | b01 / b10 | {mc['baseline_wrong_full_correct']} / {mc['baseline_correct_full_wrong']} |",
        f"| McNemar's test | two-sided exact p | {fmt_p(mc['p_value'])} |",
        f"| Paired bootstrap | Delta Top-1 | {fmt_pp(delta['top1'])} pp, 95% CI [{fmt_pp(ci['top1'][0])}, {fmt_pp(ci['top1'][1])}] |",
        f"| Paired bootstrap | Delta Top-5 | {fmt_pp(delta['top5'])} pp, 95% CI [{fmt_pp(ci['top5'][0])}, {fmt_pp(ci['top5'][1])}] |",
        f"| Paired bootstrap | Delta Macro-F1 | {fmt_pp(delta['macro_f1'])} pp, 95% CI [{fmt_pp(ci['macro_f1'][0])}, {fmt_pp(ci['macro_f1'][1])}] |",
        f"| Paired bootstrap | Delta Weighted-F1 | {fmt_pp(delta['weighted_f1'])} pp, 95% CI [{fmt_pp(ci['weighted_f1'][0])}, {fmt_pp(ci['weighted_f1'][1])}] |",
        f"| Wilcoxon signed-rank | mean class Delta F1 | {fmt_pp(summary['wilcoxon_mean_delta_f1'])} pp |",
        f"| Wilcoxon signed-rank | positive / negative / zero classes | {summary['wilcoxon_positive_classes']} / {summary['wilcoxon_negative_classes']} / {summary['wilcoxon_zero_classes']} |",
        f"| Wilcoxon signed-rank | one-sided p, Delta F1 > 0 | {fmt_p(float(wx.pvalue))} |",
    ]
    (out_dir / "significance_summary.md").write_text("\n".join(md), encoding="utf-8")
    print("\n".join(md))
    print(f"\nSaved to: {out_dir}")


if __name__ == "__main__":
    main()

File: eval/compute_significance_stats.js
const fs = require("fs");
const path = require("path");

const root = path.resolve(__dirname, "..");
const evalRoot = path.join(root, "repaet run eval");
const outDir = path.join(evalRoot, "significance_stats");
fs.mkdirSync(outDir, { recursive: true });

const seeds = [1, 25, 42, 50, 100];
const NUM_CLASSES = 241;
const BOOT = 2000;

function parseCsvLine(line) {
  const out = [];
  let cur = "";
  let inQ = false;
  for (let i = 0; i < line.length; i++) {
    const ch = line[i];
    if (ch === '"') {
      if (inQ && line[i + 1] === '"') {
        cur += '"';
        i++;
      } else inQ = !inQ;
    } else if (ch === "," && !inQ) {
      out.push(cur);
      cur = "";
    } else cur += ch;
  }
  out.push(cur);
  return out;
}

function readCsv(file) {
  const text = fs.readFileSync(file, "utf8").replace(/^\uFEFF/, "").trim();
  const lines = text.split(/\r?\n/).filter(Boolean);
  const header = parseCsvLine(lines[0]);
  return lines.slice(1).map((line) => {
    const vals = parseCsvLine(line);
    const obj = {};
    header.forEach((h, i) => {
      obj[h] = vals[i];
    });
    return obj;
  });
}

function readPred(method, seed) {
  const dir = path.join(evalRoot, method === "baseline" ? `best_baseseed${seed}` : `best_fullseed${seed}`);
  return readCsv(path.join(dir, "predictions.csv")).map((r) => ({
    path: r.path,
    y: Number(r.true_id),
    pred: Number(r.pred_id),
    top5: r.top5_ids.split(";").map(Number),
  }));
}

function readReport(method, seed) {
  const dir = path.join(evalRoot, method === "baseline" ? `best_baseseed${seed}` : `best_fullseed${seed}`);
  const rows = readCsv(path.join(dir, "classification_report.csv"));
  const f1 = new Map();
  for (const r of rows) {
    const label = r[""];
    if (/^\d+$/.test(label)) f1.set(Number(label), Number(r["f1-score"]));
  }
  return f1;
}

function metricsFromRecords(records) {
  const n = records.length;
  let top1 = 0;
  let top5 = 0;
  const tp = Array(NUM_CLASSES).fill(0);
  const fp = Array(NUM_CLASSES).fill(0);
  const fn = Array(NUM_CLASSES).fill(0);
  const support = Array(NUM_CLASSES).fill(0);
  for (const r of records) {
    if (r.pred === r.y) top1++;
    if (r.top5.includes(r.y)) top5++;
    support[r.y]++;
    if (r.pred === r.y) tp[r.y]++;
    else {
      fp[r.pred]++;
      fn[r.y]++;
    }
  }
  let macro = 0;
  let weighted = 0;
  let totalSupport = 0;
  for (let c = 0; c < NUM_CLASSES; c++) {
    const denom = 2 * tp[c] + fp[c] + fn[c];
    const f1 = denom > 0 ? (2 * tp[c]) / denom : 0;
    macro += f1;
    weighted += f1 * support[c];
    totalSupport += support[c];
  }
  return {
    top1: top1 / n,
    top5: top5 / n,
    macro_f1: macro / NUM_CLASSES,
    weighted_f1: weighted / totalSupport,
  };
}

function pairedMetrics(base, full, indices = null) {
  const b = [];
  const f = [];
  const idxs = indices || base.map((_, i) => i);
  for (const i of idxs) {
    b.push(base[i]);
    f.push(full[i]);
  }
  const mb = metricsFromRecords(b);
  const mf = metricsFromRecords(f);
  return {
    top1: mf.top1 - mb.top1,
    top5: mf.top5 - mb.top5,
    macro_f1: mf.macro_f1 - mb.macro_f1,
    weighted_f1: mf.weighted_f1 - mb.weighted_f1,
  };
}

function mulberry32(a) {
  return function rng() {
    let t = (a += 0x6d2b79f5);
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

function percentile(arr, p) {
  const a = [...arr].sort((x, y) => x - y);
  const pos = (a.length - 1) * p;
  const lo = Math.floor(pos);
  const hi = Math.ceil(pos);
  if (lo === hi) return a[lo];
  return a[lo] + (a[hi] - a[lo]) * (pos - lo);
}

function bootstrapCI(base, full, reps = BOOT, seed = 12345) {
  const rng = mulberry32(seed);
  const n = base.length;
  const vals = { top1: [], top5: [], macro_f1: [], weighted_f1: [] };
  for (let r = 0; r < reps; r++) {
    const idx = Array(n);
    for (let i = 0; i < n; i++) idx[i] = Math.floor(rng() * n);
    const d = pairedMetrics(base, full, idx);
    for (const k of Object.keys(vals)) vals[k].push(d[k]);
  }
  const out = {};
  for (const k of Object.keys(vals)) out[k] = [percentile(vals[k], 0.025), percentile(vals[k], 0.975)];
  return out;
}

function logChoose(n, k) {
  if (k < 0 || k > n) return -Infinity;
  k = Math.min(k, n - k);
  let s = 0;
  for (let i = 1; i <= k; i++) s += Math.log(n - k + i) - Math.log(i);
  return s;
}

function binomCdf(k, n, p = 0.5) {
  if (k < 0) return 0;
  if (k >= n) return 1;
  const logs = [];
  let max = -Infinity;
  for (let i = 0; i <= k; i++) {
    const v = logChoose(n, i) + i * Math.log(p) + (n - i) * Math.log(1 - p);
    logs.push(v);
    if (v > max) max = v;
  }
  let sum = 0;
  for (const v of logs) sum += Math.exp(v - max);
  return Math.exp(max) * sum;
}

function mcnemar(base, full) {
  let b01 = 0;
  let b10 = 0;
  for (let i = 0; i < base.length; i++) {
    const bc = base[i].pred === base[i].y;
    const fc = full[i].pred === full[i].y;
    if (!bc && fc) b01++;
    if (bc && !fc) b10++;
  }
  const n = b01 + b10;
  const p = Math.min(1, 2 * binomCdf(Math.min(b01, b10), n, 0.5));
  return { baseline_wrong_full_correct: b01, baseline_correct_full_wrong: b10, discordant: n, p_value: p };
}

function erf(x) {
  const sign = x < 0 ? -1 : 1;
  x = Math.abs(x);
  const a1 = 0.254829592;
  const a2 = -0.284496736;
  const a3 = 1.421413741;
  const a4 = -1.453152027;
  const a5 = 1.061405429;
  const p = 0.3275911;
  const t = 1 / (1 + p * x);
  const y = 1 - (((((a5 * t + a4) * t + a3) * t + a2) * t + a1) * t * Math.exp(-x * x));
  return sign * y;
}

function normCdf(z) {
  return 0.5 * (1 + erf(z / Math.SQRT2));
}

function wilcoxonGreater(diffs) {
  const nz = diffs.filter((x) => Math.abs(x) > 1e-15);
  const arr = nz.map((x) => ({ abs: Math.abs(x), sign: Math.sign(x) })).sort((a, b) => a.abs - b.abs);
  let W = 0;
  const tieCounts = [];
  for (let i = 0; i < arr.length;) {
    let j = i + 1;
    while (j < arr.length && Math.abs(arr[j].abs - arr[i].abs) < 1e-15) j++;
    const avg = (i + 1 + j) / 2;
    const count = j - i;
    tieCounts.push(count);
    for (let k = i; k < j; k++) if (arr[k].sign > 0) W += avg;
    i = j;
  }
  const n = arr.length;
  const mean = (n * (n + 1)) / 4;
  let variance = (n * (n + 1) * (2 * n + 1)) / 24;
  let tieAdj = 0;
  for (const t of tieCounts) tieAdj += t * t * t - t;
  variance -= tieAdj / 48;
  const z = (W - mean - 0.5) / Math.sqrt(variance);
  const p = 1 - normCdf(z);
  return {
    n_nonzero: n,
    W_plus: W,
    z,
    p_value_greater: p,
    mean_delta: diffs.reduce((a, b) => a + b, 0) / diffs.length,
    positives: diffs.filter((x) => x > 0).length,
    negatives: diffs.filter((x) => x < 0).length,
    zeros: diffs.filter((x) => Math.abs(x) <= 1e-15).length,
  };
}

function csvEscape(x) {
  const s = String(x);
  return /[",\n]/.test(s) ? `"${s.replace(/"/g, '""')}"` : s;
}

function writeCsv(file, rows) {
  const keys = Object.keys(rows[0]);
  const lines = [keys.join(",")];
  for (const r of rows) lines.push(keys.map((k) => csvEscape(r[k])).join(","));
  fs.writeFileSync(file, lines.join("\n"), "utf8");
}

function fmtPct(x) {
  return (x * 100).toFixed(3);
}

function fmtP(p) {
  return p < 1e-4 ? p.toExponential(3) : p.toFixed(6);
}

const perSeedRows = [];
const pooledBase = [];
const pooledFull = [];

for (const seed of seeds) {
  const b = readPred("baseline", seed);
  const f = readPred("full", seed);
  if (b.length !== f.length) throw new Error(`length mismatch seed ${seed}`);
  for (let i = 0; i < b.length; i++) {
    if (b[i].path !== f[i].path || b[i].y !== f[i].y) throw new Error(`pair mismatch seed ${seed} idx ${i}`);
  }
  pooledBase.push(...b);
  pooledFull.push(...f);
  const d = pairedMetrics(b, f);
  const ci = bootstrapCI(b, f, BOOT, 1000 + seed);
  const mc = mcnemar(b, f);
  const rb = readReport("baseline", seed);
  const rf = readReport("full", seed);
  const diffs = [];
  for (let c = 0; c < NUM_CLASSES; c++) diffs.push((rf.get(c) || 0) - (rb.get(c) || 0));
  const wx = wilcoxonGreater(diffs);
  perSeedRows.push({
    seed,
    n: b.length,
    delta_top1: d.top1,
    delta_top5: d.top5,
    delta_macro_f1: d.macro_f1,
    delta_weighted_f1: d.weighted_f1,
    top1_ci_low: ci.top1[0],
    top1_ci_high: ci.top1[1],
    top5_ci_low: ci.top5[0],
    top5_ci_high: ci.top5[1],
    macro_f1_ci_low: ci.macro_f1[0],
    macro_f1_ci_high: ci.macro_f1[1],
    weighted_f1_ci_low: ci.weighted_f1[0],
    weighted_f1_ci_high: ci.weighted_f1[1],
    mcnemar_b01: mc.baseline_wrong_full_correct,
    mcnemar_b10: mc.baseline_correct_full_wrong,
    mcnemar_p: mc.p_value,
    wilcoxon_W_plus: wx.W_plus,
    wilcoxon_z: wx.z,
    wilcoxon_p_greater: wx.p_value_greater,
    wilcoxon_mean_delta_f1: wx.mean_delta,
    wilcoxon_pos: wx.positives,
    wilcoxon_neg: wx.negatives,
    wilcoxon_zero: wx.zeros,
  });
}

const pooledD = pairedMetrics(pooledBase, pooledFull);
const pooledCI = bootstrapCI(pooledBase, pooledFull, BOOT, 999);
const pooledMC = mcnemar(pooledBase, pooledFull);
const meanClassDiffs = [];
for (let c = 0; c < NUM_CLASSES; c++) {
  let s = 0;
  for (const seed of seeds) {
    const rb = readReport("baseline", seed);
    const rf = readReport("full", seed);
    s += (rf.get(c) || 0) - (rb.get(c) || 0);
  }
  meanClassDiffs.push(s / seeds.length);
}
const pooledWX = wilcoxonGreater(meanClassDiffs);

writeCsv(path.join(outDir, "significance_per_seed.csv"), perSeedRows);
writeCsv(path.join(outDir, "mean_class_delta_f1.csv"), meanClassDiffs.map((d, i) => ({ class_id: i, mean_delta_f1: d })));

const summaryRows = [{
  comparison: "Full model vs Baseline (pooled seed-image pairs)",
  seeds: seeds.join(";"),
  n_pairs: pooledBase.length,
  mcnemar_b01: pooledMC.baseline_wrong_full_correct,
  mcnemar_b10: pooledMC.baseline_correct_full_wrong,
  mcnemar_discordant: pooledMC.discordant,
  mcnemar_p_value: pooledMC.p_value,
  delta_top1: pooledD.top1,
  delta_top1_ci: `[${pooledCI.top1[0]}, ${pooledCI.top1[1]}]`,
  delta_top5: pooledD.top5,
  delta_top5_ci: `[${pooledCI.top5[0]}, ${pooledCI.top5[1]}]`,
  delta_macro_f1: pooledD.macro_f1,
  delta_macro_f1_ci: `[${pooledCI.macro_f1[0]}, ${pooledCI.macro_f1[1]}]`,
  delta_weighted_f1: pooledD.weighted_f1,
  delta_weighted_f1_ci: `[${pooledCI.weighted_f1[0]}, ${pooledCI.weighted_f1[1]}]`,
  wilcoxon_n_classes_nonzero: pooledWX.n_nonzero,
  wilcoxon_W_plus: pooledWX.W_plus,
  wilcoxon_z: pooledWX.z,
  wilcoxon_p_greater: pooledWX.p_value_greater,
  wilcoxon_mean_delta_f1: pooledWX.mean_delta,
  wilcoxon_pos_classes: pooledWX.positives,
  wilcoxon_neg_classes: pooledWX.negatives,
  wilcoxon_zero_classes: pooledWX.zeros,
}];
writeCsv(path.join(outDir, "significance_summary.csv"), summaryRows);

const md = [];
md.push("| Analysis | Statistic | Result |");
md.push("|---|---:|---:|");
md.push(`| McNemar's test | b01 / b10 | ${pooledMC.baseline_wrong_full_correct} / ${pooledMC.baseline_correct_full_wrong} |`);
md.push(`| McNemar's test | two-sided exact p | ${fmtP(pooledMC.p_value)} |`);
md.push(`| Paired bootstrap | ΔTop-1 | ${fmtPct(pooledD.top1)} pp, 95% CI [${fmtPct(pooledCI.top1[0])}, ${fmtPct(pooledCI.top1[1])}] |`);
md.push(`| Paired bootstrap | ΔTop-5 | ${fmtPct(pooledD.top5)} pp, 95% CI [${fmtPct(pooledCI.top5[0])}, ${fmtPct(pooledCI.top5[1])}] |`);
md.push(`| Paired bootstrap | ΔMacro-F1 | ${fmtPct(pooledD.macro_f1)} pp, 95% CI [${fmtPct(pooledCI.macro_f1[0])}, ${fmtPct(pooledCI.macro_f1[1])}] |`);
md.push(`| Paired bootstrap | ΔWeighted-F1 | ${fmtPct(pooledD.weighted_f1)} pp, 95% CI [${fmtPct(pooledCI.weighted_f1[0])}, ${fmtPct(pooledCI.weighted_f1[1])}] |`);
md.push(`| Wilcoxon signed-rank | mean class ΔF1 | ${fmtPct(pooledWX.mean_delta)} pp |`);
md.push(`| Wilcoxon signed-rank | positive / negative / zero classes | ${pooledWX.positives} / ${pooledWX.negatives} / ${pooledWX.zeros} |`);
md.push(`| Wilcoxon signed-rank | one-sided p, ΔF1 > 0 | ${fmtP(pooledWX.p_value_greater)} |`);
fs.writeFileSync(path.join(outDir, "significance_summary.md"), md.join("\n"), "utf8");

console.log(md.join("\n"));
console.log(`Saved to ${outDir}`);
